### PR TITLE
Geanina/Create all classes and added first functionality: add an employee with the ability to add data in the file without overwriting the existing data

### PR DESCRIPTION
**Pull request by finarugeanina**
- all classes needed for the project were added
- only the option 1. Add a new employee is functional

### CODE REVIEW

File: src/EmployeesAdministrationApp.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+			convertEmployee();
+		case 7:
+			System.out.println("Bye Bye!");
+
+		}
+		scan.close();
+	}
+
+	private static void addNewEmployee() throws Exception {
+		{
```
Comment (rbutnariu) [inline src/EmployeesAdministrationApp.java]: why is this needed?

File: src/EmployeesAdministrationApp.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+			searchEmployee();
+			break;
+		case 4:
+			deleteEmployee();
+			break;
+		case 5:
+			updateEmployee();
+			break;
+		case 6:
+			convertEmployee();
```
Comment (rbutnariu) [inline src/EmployeesAdministrationApp.java:35]: why don't we have break here?

File: src/EmployeesAdministrationApp.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+		case 7:
+			System.out.println("Bye Bye!");
+
+		}
+		scan.close();
+	}
+
+	private static void addNewEmployee() throws Exception {
+		{
+			System.out.println(
```
Comment (rbutnariu) [inline src/EmployeesAdministrationApp.java]: I think this could be better formatted.

File: src/Employee.java [hunk 4]
```diff
@@ -1,14 +1,110 @@
...
+	public String getCNP() {
+		return CNP;
+	}
+
+	/**
+	 * This sets the CNP for the employee
+	 * 
+	 * @param cNP
+	 */
+	public void setCNP(String cNP) {
```
Comment (rbutnariu) [inline src/Employee.java]: typo cNP

File: src/Employee.java [hunk 5]
```diff
@@ -1,14 +1,110 @@
...
+ * 
+ * @author geanina.finaru
+ * @version 1.0
+ */
+public abstract class Employee {
 
 	private String name;
+	private int employeeId;
+	private String gender;
+	private String CNP;
```
Comment (rbutnariu) [inline src/Employee.java:22]: let's add some validation for this attribute. (i.e 13 characters with first character being 1 or 2, next 6 digits - a valid date in format YYmmDD, etc)

File: src/DeveloperEmployee.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+	}
+
+	@Override
+	public String toString() {
+		return super.toString() + String.format("\nGeneral level of Developer Knowings : %s", devGeneralLevel);
+
+	}
+
+	@SuppressWarnings("unchecked")
```
Comment (rbutnariu) [inline src/DeveloperEmployee.java]: why is this needed?

File: src/DeveloperEmployee.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Scanner;
+
+import org.json.simple.JSONObject;
+
+public class DeveloperEmployee extends Employee {
+	private int devGeneralLevel;
+	public final static String title = "DEV";
```
Comment (rbutnariu) [inline src/DeveloperEmployee.java]: why is this needed?

File: src/ProjectManager.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Scanner;
+
+import org.json.simple.JSONObject;
+
+public class ProjectManager extends Employee {
+	protected int pmExperience;
+	public final static String title = "PM";
```
Comment (rbutnariu) [inline src/ProjectManager.java]: is this needed?

**Review by rbutnariu — COMMENTED** [review]
there are a lot of similar methods with the only difference being the employee type. think of a method to reduce the amount of duplicated code. try to parameterize the methods where posibile and check if the "reflection" can be used https://www.geeksforgeeks.org/reflection-in-java/

**Comment by rbutnariu** [comment]
avoid having too many empty lines between methods. follow java best practices

File: src/DeveloperEmployee.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+	@SuppressWarnings("unchecked")
+	public void writeJSONFileForDevEmployee(DeveloperEmployee dev, String filename) throws Exception {
+		JSONObject employeeDetails = new JSONObject();
+		employeeDetails.put("Name", dev.getName());
+		employeeDetails.put("Employee Id", dev.getEmployeeId());
+		employeeDetails.put("CNP", dev.getCNP());
+		employeeDetails.put("Gender", dev.getGender());
+		employeeDetails.put("General Developer Level", dev.getDevGeneralLevel());
+		System.out
+				.println(
```
Comment (rbutnariu) [inline src/DeveloperEmployee.java]: keep this on a single line

**Comment by rbutnariu** [comment]
Please update the title and make it more specific and clear

File: src/EmployeesAdministrationApp.java (hunk 1 above)
Comment (finarugeanina) [reply src/EmployeesAdministrationApp.java]: I added a generic class which uses the type of the created instance as parameter for avoiding duplicated methods.

Comment (finarugeanina) [reply src/EmployeesAdministrationApp.java]: I removed "thrown Exception" as it was a failed trial of resolving a exception :)

File: src/EmployeesAdministrationApp.java (hunk 2 above)
Comment (finarugeanina) [reply src/EmployeesAdministrationApp.java:35]: The break is added now

File: src/Employee.java (hunk 5 above)
Comment (finarugeanina) [reply src/Employee.java:22]: I added the validation criteria for this attribute

File: src/DeveloperEmployee.java (hunk 6 above)
Comment (finarugeanina) [reply src/DeveloperEmployee.java]: Apparently I had a warning for the JSONObject and I didn't know how to resolve it for the moment, so this was for not reminding me about it :) . It's resolved now (I used a hashmap and then I created the JSONObject with this hashmap as a parameter).

File: src/DeveloperEmployee.java (hunk 7 above)
Comment (finarugeanina) [reply src/DeveloperEmployee.java]: I thought I will implement my generic class which uses this as a parameter, but it was a wrong idea :) I delete it for all.

File: src/DeveloperEmployee.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import java.util.Scanner;
+
+public class DeveloperEmployee extends Employee {
+	private int devGeneralLevel;
```
Comment (rbutnariu) [inline src/DeveloperEmployee.java:4]: add some validation here (i.e. from 1 to 10)

File: src/DeveloperTeamLead.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
+import java.util.Scanner;
+
+public class DeveloperTeamLead extends DeveloperEmployee {
+	private int developerExperience;
```
Comment (rbutnariu) [inline src/DeveloperTeamLead.java:4]: add some validation here

File: src/Employee.java [hunk 12]
```diff
@@ -1,14 +1,139 @@
...
 
+/**
+ * This class implements a simple employee
+ * 
+ * @author geanina.finaru
+ * @version 1.0
+ */
 public class Employee {
 
 	private String name;
```
Comment (rbutnariu) [inline src/Employee.java:19]: add some validation here

File: src/Employee.java [hunk 13]
```diff
@@ -1,14 +1,139 @@
...
-
+	private int employeeId;
```
Comment (rbutnariu) [inline src/Employee.java:20]: add some validation here (the id must be unique)

File: src/Employee.java [hunk 14]
```diff
@@ -1,14 +1,139 @@
...
+	public void setCNP(String CNP) {
+		if (CNP.length() != 13) {
+			System.out.println("The length of the CNP should be 13 numbers! Please re-enter the CNP:");
+			setCNP(scan.nextLine());
+
+		} else if (!CNP.substring(0, 1).equals("1") && !CNP.substring(0, 1).equals("2")) {
+			System.out.println("The CNP should start only with 1 or 2! Please re-enter the CNP:");
+			setCNP(scan.nextLine());
+		} else if (!isValidDateFormat(CNP.substring(1, 7))) {
+			System.out.println("The enetered CNP is not in the right date format! Please re-enter the CNP:)");
```
Comment (rbutnariu) [inline src/Employee.java]: type "entered"

File: src/QATeamLead.java [hunk 15]
```diff
@@ -0,0 +1,28 @@
+import java.util.Scanner;
+
+public class QATeamLead extends QAEmployee {
+	private int qaExperience;
```
Comment (rbutnariu) [inline src/QATeamLead.java:4]: add some validation here

File: src/QAEmployee.java [hunk 16]
```diff
@@ -0,0 +1,31 @@
+import java.util.Scanner;
+
+public class QAEmployee extends Employee {
+	private int qaGeneralLevel;
```
Comment (rbutnariu) [inline src/QAEmployee.java:4]: add some validation here

File: src/ProjectManager.java [hunk 17]
```diff
@@ -0,0 +1,32 @@
+import java.util.Scanner;
+
+public class ProjectManager extends Employee {
+	protected int pmExperience;
```
Comment (rbutnariu) [inline src/ProjectManager.java:4]: add some validation here

File: src/EmployeesAdministrationApp.java [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+				writeJson(projectManager);
+				break;
+			}
+			}
+			scan.close();
+	}
+
+	private static void writeJson(Object obj) {
+		Method[] methods = obj.getClass().getMethods();
+		HashMap<String, Object> employeeDetails = new HashMap<String, Object>();
```
Comment (rbutnariu) [inline src/EmployeesAdministrationApp.java]: use Map when declaring

Comment (rbutnariu) [reply src/EmployeesAdministrationApp.java]: Map<String, Object> employeeDetails = new HashMap<String, Object>();

File: src/EmployeesAdministrationApp.java [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+		for (Method method : methods) {
+			if (method.getName().startsWith("get") && !method.getName().equals("getClass")) {
```
Comment (rbutnariu) [inline src/EmployeesAdministrationApp.java]: what if you will have another method that starts with get besides getClass? currently, all the methods that start with "get" will be invoked

Comment (rbutnariu) [reply src/EmployeesAdministrationApp.java]: maybe this will help https://www.baeldung.com/java-method-reflection

File: src/EmployeesAdministrationApp.java [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+		System.out.println("Please select the type of the new employee:"
+				+ "\n1. QA Employee"
+				+ "\n2. Developer Employee"
+				+ "\n3. QA Team Lead Employee " + "\n4. Developer Team Lead Employee" + "\n5. Project Manager");
+			Scanner scan = new Scanner(System.in);
+			int typeOfTheEmployee = scan.nextInt();
+
+			switch (typeOfTheEmployee) {
+			case 1: {
+				QAEmployee qa = new QAEmployee();
```
Comment (rbutnariu) [inline src/EmployeesAdministrationApp.java]: I noticed all the code from the corresponding method has been moved here. this is still duplicated (write a general method that takes as parameters the string to print and the object type as Object. you can check what class instance an object is by using instanceof)
E.g: void generalMethod(String print, Object employee) {<>}
call: generalMethod("Enter the QA employee data", qa);

Comment (finarugeanina) [reply src/EmployeesAdministrationApp.java]: I added a general method which uses the message and the new instance of an class as parameters.

File: src/EmployeesAdministrationApp.java (hunk 19 above)
Comment (finarugeanina) [reply src/EmployeesAdministrationApp.java]: all methods which are not named "getClass" should be invoked and the result added in the JSONObject.

File: src/EmployeesAdministrationApp.java (hunk 18 above)
Comment (finarugeanina) [reply src/EmployeesAdministrationApp.java]: I added the correct declaration.

File: src/ProjectManager.java (hunk 17 above)
Comment (finarugeanina) [reply src/ProjectManager.java:4]: I added the validation for this attribute.

File: src/QAEmployee.java (hunk 16 above)
Comment (finarugeanina) [reply src/QAEmployee.java:4]: I added the validation for this attribute.

File: src/QATeamLead.java (hunk 15 above)
Comment (finarugeanina) [reply src/QATeamLead.java:4]: I added the validation for this attribute.

File: src/Employee.java (hunk 14 above)
Comment (finarugeanina) [reply src/Employee.java]: I modified the typo

File: src/Employee.java (hunk 13 above)
Comment (finarugeanina) [reply src/Employee.java:20]: I added the validation for this attribute.

File: src/Employee.java (hunk 12 above)
Comment (finarugeanina) [reply src/Employee.java:19]: I added the validation for this attribute.

File: src/DeveloperTeamLead.java (hunk 11 above)
Comment (finarugeanina) [reply src/DeveloperTeamLead.java:4]: I added the validation for this attribute.

File: src/DeveloperEmployee.java (hunk 10 above)
Comment (finarugeanina) [reply src/DeveloperEmployee.java:4]: I added the validation for this attribute.

File: src/DeveloperEmployee.java (hunk 9 above)
Comment (finarugeanina) [reply src/DeveloperEmployee.java]: This is changed now

File: src/ProjectManager.java (hunk 8 above)
Comment (finarugeanina) [reply src/ProjectManager.java]: I thought I will implement a generic class which uses this as a parameter, but it was a wrong idea :) I delete it for all.

File: src/Employee.java (hunk 4 above)
Comment (finarugeanina) [reply src/Employee.java]: I changed the typo

File: src/EmployeesAdministrationApp.java (hunk 3 above)
Comment (finarugeanina) [reply src/EmployeesAdministrationApp.java]: This is changed now.

**Comment by rbutnariu** [comment]
👍

**Comment by rbutnariu** [comment]
looks good to me. merging

**Review by rbutnariu — APPROVED** [review]
